tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py

```diff
@@ -0,0 +1,770 @@
+                    market_state=MarketState(
+                        share_reserves=100000,  # base reserves (in share terms) base = share * share_price
+                        bond_reserves=1000000,  # PT reserves
+                        share_price=2,  # share price of the LP in the yield source
+                        init_share_price=1.5,  # original share price pool started
+                    ),
+                    fee_percent=0.1,  # fee percent (normally 10%)
+                    days_remaining=91.25,  # 3.0 months remaining
+                    time_stretch_apy=0.1,  # APY used to calculate time_stretch
+                    test_id=8,
+                ),
+                # in_base=189.25267474842258
+                # in_pt=213.6913557597436
+                TestResultCalcInGivenOutSuccess(
+                    without_fee_or_slippage=189.25228759556148,
+                    without_fee=189.25267474842258,
+                    hyperdrive_fee=1.0747712404438503,
+                    hyperdrive_with_fee=190.32744598886643,
+                ),
+            ),  # end of test 16
+        ]
+
+        test_cases = base_in_test_cases + pt_in_test_cases
+
+        pricing_model = HyperdrivePricingModel()
+        for (
+            test_case,
+            expected_result,
+        ) in test_cases:
+            time_stretch = pricing_model.calc_time_stretch(test_case.time_stretch_apy)
+            time_remaining = StretchedTime(days=test_case.days_remaining, time_stretch=time_stretch)
+
+            # Ensure we get the expected results from the pricing model.
+            trade_result = pricing_model.calc_in_given_out(
+                out=test_case.out,
+                market_state=test_case.market_state,
+                fee_percent=test_case.fee_percent,
+                time_remaining=time_remaining,
+            )
+
+            np.testing.assert_almost_equal(
+                trade_result.breakdown.without_fee_or_slippage,
+                expected_result.without_fee_or_slippage,
+                err_msg=f"test {test_case.test_id} unexpected without_fee_or_slippage",
+            )
+            np.testing.assert_almost_equal(
+                trade_result.breakdown.without_fee,
+                expected_result.without_fee,
+                err_msg=f"test {test_case.test_id} unexpected without_fee",
+            )
+            np.testing.assert_almost_equal(
+                trade_result.breakdown.fee,
+                expected_result.hyperdrive_fee,
+                err_msg=f"test {test_case.test_id} unexpected hyperdrive fee",
+            )
+            np.testing.assert_almost_equal(
+                trade_result.breakdown.with_fee,
+                expected_result.hyperdrive_with_fee,
+                err_msg=f"test {test_case.test_id} unexpected hyperdrive with_fee",
+            )
+
+    # # TODO: This should be refactored to be a test for check_input_assertions and check_output_assertions
```

Comment:
This will need to be uncommented before it's merged.

Reply:
Added failure tests